adapters/pubmatic/pubmatic.go

```diff
@@ -320,31 +323,96 @@ func parseImpressionObject(imp *openrtb2.Imp, extractWrapperExtFromImp, extractP
 }
 
 // extractPubmaticExtFromRequest parse the req.ext to fetch wrapper and acat params
-func extractPubmaticExtFromRequest(request *openrtb2.BidRequest) (*pubmaticWrapperExt, []string, error) {
-	var acat []string
-	var wrpExt *pubmaticWrapperExt
+func extractPubmaticExtFromRequest(request *openrtb2.BidRequest) (extRequestAdServer, error) {
+	// req.ext.prebid would always be there and Less nil cases to handle, more safe!
+	var pmReqExt extRequestAdServer
+
+	if request == nil || len(request.Ext) == 0 {
+		return pmReqExt, nil
+	}
+
+	// todo: unvoid repeated unmarshal of request.ext
+	reqExt := &openrtb_ext.ExtRequest{}
+	err := json.Unmarshal(request.Ext, &reqExt)
+	if err != nil {
+		return pmReqExt, fmt.Errorf("error decoding Request.ext : %s", err.Error())
+	}
+	pmReqExt.ExtRequest = *reqExt
+
 	reqExtBidderParams, err := adapters.ExtractReqExtBidderParamsMap(request)
 	if err != nil {
-		return nil, acat, err
+		return pmReqExt, err
 	}
 
 	//get request ext bidder params
 	if wrapperObj, present := reqExtBidderParams["wrapper"]; present && len(wrapperObj) != 0 {
-		wrpExt = &pubmaticWrapperExt{}
+		wrpExt := &pubmaticWrapperExt{}
 		err = json.Unmarshal(wrapperObj, wrpExt)
 		if err != nil {
-			return nil, acat, err
+			return pmReqExt, err
 		}
+		pmReqExt.Wrapper = wrpExt
 	}
 
 	if acatBytes, ok := reqExtBidderParams["acat"]; ok {
+		var acat []string
 		err = json.Unmarshal(acatBytes, &acat)
+		if err != nil {
+			return pmReqExt, err
+		}
 		for i := 0; i < len(acat); i++ {
 			acat[i] = strings.TrimSpace(acat[i])
 		}
+		pmReqExt.Acat = acat
 	}
 
-	return wrpExt, acat, err
+	if allowedBidders := getAlternateBidderCodesFromRequestExt(reqExt); allowedBidders != nil {
+		pmReqExt.Marketplace = &marketplaceReqExt{AllowedBidders: allowedBidders}
+	}
+
+	return pmReqExt, nil
+}
+
+func getAlternateBidderCodesFromRequestExt(reqExt *openrtb_ext.ExtRequest) []string {
+	if reqExt == nil || reqExt.Prebid.AlternateBidderCodes == nil {
+		return nil
+	}
+
+	allowedBidders := []string{"pubmatic"}
+	if reqExt.Prebid.AlternateBidderCodes.Enabled {
+		if pmABC, ok := reqExt.Prebid.AlternateBidderCodes.Bidders["pubmatic"]; ok && pmABC.Enabled {
+			if pmABC.AllowedBidderCodes == nil || (len(pmABC.AllowedBidderCodes) == 1 && pmABC.AllowedBidderCodes[0] == "*") {
+				return []string{"all"}
+			}
+			// keeping this off for now. Makes more sense to move this to PBS-Core
+			// return append(allowedBidders, assertBidderCodes(pmABC.AllowedBidderCodes)...)
+			return append(allowedBidders, pmABC.AllowedBidderCodes...)
+		}
+	}
+
+	return allowedBidders
+}
+
+// assertBidderCodes typical string assertions: remove duplicates, remove whitespaces, all lowercase case biddercode
+func assertBidderCodes(bidders []string) []string {
+	// nil not expected as it is a special case to allow all bidders
+	if bidders == nil {
+		return nil
+	}
+
+	checkedBidders := map[string]bool{"pubmatic": true}
```

Comment:
Should `checkedBidders` just be an empty map? It appears that when you call this you would have already added `"pubmatic"` to the bidders array so it would just be handled here when you loop over the bidders.

Reply:
removed the comment w.r.t above comments